File: ruby/matrix/matrix.rb
class Matrix
  def initialize(input)
    rows = input.split("\n")
    @matrix = rows.map do |r|
      r.split(' ').map(&:to_i)
    end
  end

  def rows
    @matrix
  end

  def columns
    0.upto(@matrix.length).map do |i|
      @matrix.map { |r| r[i] }
    end
  end
end

File: scala/matrix/src/main/scala/Matrix.scala
case class Matrix(raw: String) {
  val data = raw
    .split("\n")
    .map(row => row.split(" ").map(_.toInt).toVector)
    .toVector

  def rows = data
  def cols(idx: Int) = data.map(row => row(idx))

  override def equals(obj: scala.Any): Boolean =
    obj.isInstanceOf[Matrix] && obj.asInstanceOf[Matrix].data == this.data
}
